swarm/swarm.go

```diff
@@ -0,0 +1,382 @@
+	}
+
+	if o.KubernetesOptions.LabelSelectorValue == "" {
+		log.Errorf("LabelSelectorValue is empty, set to default %s instead", DefaultLabelSelectorValue)
+		o.KubernetesOptions.LabelSelectorValue = DefaultLabelSelectorValue
+	}
+
+	if o.MaxMessageBuffer <= 0 {
+		log.Errorf("MaxMessageBuffer <= 0, setting to default %d instead", DefaultMaxMessageBuffer)
+		o.MaxMessageBuffer = DefaultMaxMessageBuffer
+	}
+
+	if o.LeaveTimeout <= 0 {
+		log.Errorf("LeaveTimeout <= 0, setting to default %d instead", DefaultLeaveTimeout)
+		o.LeaveTimeout = DefaultLeaveTimeout
+	}
+
+	return Start(o)
+}
+
+// Start will find Swarm peers based on the chosen swarm type and join
+// the Swarm.
+func Start(o Options) (*Swarm, error) {
+	knownEntryPoint, cleanupF := newKnownEntryPoint(o)
+	log.Debugf("knownEntryPoint: %s, %v", knownEntryPoint.Node(), knownEntryPoint.Nodes())
+	return Join(o, knownEntryPoint.Node(), knownEntryPoint.Nodes(), cleanupF)
+}
+
+// Join will join given Swarm peers and return an initialiazed Swarm
+// object if successful.
+func Join(o Options, self *NodeInfo, nodes []*NodeInfo, cleanupF func()) (*Swarm, error) {
+	log.Infof("SWARM: %s is going to join swarm of %d nodes (%v)", self, len(nodes), nodes)
+	cfg := memberlist.DefaultLocalConfig()
+	if !o.Debug {
+		cfg.LogOutput = ioutil.Discard
+	}
+
+	if self.Name == "" {
+		self.Name = cfg.Name
+	} else {
+		cfg.Name = self.Name
+	}
+	if self.Addr == nil {
+		self.Addr = net.ParseIP(cfg.BindAddr)
+	} else {
+		cfg.BindAddr = self.Addr.String()
+		cfg.AdvertiseAddr = cfg.BindAddr
+	}
+	if self.Port == 0 {
+		self.Port = uint16(cfg.BindPort)
+	} else {
+		cfg.BindPort = int(self.Port)
+		cfg.AdvertisePort = cfg.BindPort
+	}
+
+	getOutgoing := make(chan reqOutgoing)
+	outgoing := make(chan *outgoingMessage)
+	incoming := make(chan []byte)
+	getValues := make(chan *valueReq)
+	listeners := make(map[string]chan<- *Message)
+	leave := make(chan struct{})
+	shared := make(sharedValues)
+
+	cfg.Delegate = &mlDelegate{
+		outgoing: getOutgoing,
+		incoming: incoming,
+	}
+	ml, err := memberlist.Create(cfg)
+	if err != nil {
+		log.Errorf("SWARM: failed to create memberlist: %v", err)
+		return nil, err
+	}
+	cfg.Delegate.(*mlDelegate).meta = ml.LocalNode().Meta
+
+	if len(nodes) > 0 {
+		addresses := mapNodesToAddresses(nodes)
+		_, err := ml.Join(addresses)
+		if err != nil {
+			log.Errorf("SWARM: failed to join: %v", err)
+			return nil, err
+		}
+	}
+
+	s := &Swarm{
+		local:            self,
+		maxMessageBuffer: o.MaxMessageBuffer,
+		leaveTimeout:     o.LeaveTimeout,
+		getOutgoing:      getOutgoing,
+		outgoing:         outgoing,
+		incoming:         incoming,
+		getValues:        getValues,
+		listeners:        listeners,
+		leave:            leave,
+		shared:           shared,
+		mlist:            ml,
+		cleanupF:         cleanupF,
+		metrics:          metrics.Default,
+	}
+
+	// TODO(sszuecs): maybe we should wrap it in a recover for panic, but we need to close the channels
```

Comment:
what would be the recovery strategy in case of panic?

we don't need to close the channels, but we need to return from the goroutine. This can be done if we call the Leave() function. Suggested in skipper.go

Reply:
I dropped the TODO and did the Leave as suggested.
Something open here?

Reply:
lgtm